### ISSUE TITLE
Query generated using EnableDateRangeOptimization optimized ELM does not work

### ISSUE THREAD

**Issue by csandersdev** [DBCG/cql_engine#467]
I've been trying to get the CQL below working with the EnableDateRangeOptimization feature in the translator and the RestFhirRetrieveProvider. I can't get it to pass dateTime parameters to the FHIR search call. What's worse, it does so silently and the search returns data that doesn't match the CQL author's intention. I've tried this with the 1.5.3 translator and also 1.5.4-SNAPSHOT and the behavior is the same.

```
library "UsesDateFilter" version '1.0.0'

using FHIR version '4.0.0'
include "FHIRHelpers" version '4.0.0' called FHIRHelpers
parameter "MeasurementPeriod" Interval<DateTime>
context Patient
define ConditionInInterval:
        [Condition] c where c.onset.as(dateTime) during "MeasurementPeriod"
```

This compiles to the following expression...

```xml
         <expression localId="14" locator="8:9-8:75" xsi:type="Query">
            <source localId="7" locator="8:9-8:21" alias="c">
               <expression localId="6" locator="8:9-8:19" dataType="fhir:Condition" dateProperty="onset" xsi:type="Retrieve">
                  <dateRange localId="12" locator="8:57-8:75" name="MeasurementPeriod" xsi:type="ParameterRef"/>
               </expression>
            </source>
         </expression>
```

Which generates a retrieve call that looks like the following...

```java
provider.retrieve("Patient", "subject", "123", "Condition", null, null, null, null, "onset", null, null, interval);
```

The datePath parameter "onset" never matches a REST search parameter definition (e.g. onset-date), so the search parameter resolution just passes over the value. I would be happy to try to fix this, but there are a lot of moving parts between the modelinfo, the translator, and the engine and I'm not sure exactly where to start figuring this out.

**Comment by JPercival** [comment]
There are gaps in the cql-engine and/or data provider implementations around converting filters to search parameters. In general, it's not guaranteed that a given codePath, datePath, etc. (which are all passed to the DataProvider in terms of paths on the model) will be supported by any search parameter on a given FHIR server (which may or may not map directly to the path on the model - e.g. `Observation` using `patient=123` for the search parameter, while `subject` is the path). There may also be multiple search parameters that map to one path or a server may not support any search parameters at all. We've largely sidestepped this issue so far because the cqf-ruler supports most of the search parameters defined by the FHIR spec.

There are at least two possible (not mutually exclusive) fixes for this:

1. The cql-engine reads the capabilities of the target server and rewrites the ELM accordingly, adding in filter ELM nodes where needed to get the correct results in the case the target server

2. We make the DataProviders fully implement all the possible parameters for a Retrieve. IOW, if a FHIR server doesn't support a given datePath, the DataProvider needs to apply additional filtering client-side.

There's a bit of work in the cql-evaluator project towards that second option since it can work against a filesystem directly, which obviously doesn't support any search parameters, indexes, etc. Rolling that into the engine wouldn't be particularly difficult but I've demurred since there's the potential for unintended side-effects and we've lacked testing around the FHIR-based retrieve providers.

**Comment by JPercival** [comment]
An example of "client-side" filtering:

https://github.com/DBCG/cql-evaluator/blob/master/evaluator.engine/src/main/java/org/opencds/cqf/cql/evaluator/engine/retrieve/BundleRetrieveProvider.java#L119

**Comment by JPercival** [comment]
Issue tracking some of the query planning work:

https://github.com/cqframework/cqf-tooling/issues/220

**Comment by JPercival** [comment]
And somehow, the date filter support made it into the roadmap, but we don't have an issue for it:

https://github.com/DBCG/cql_engine/blob/master/README.md#roadmap

So this is the one!

**Comment by csandersdev** [comment]
Thanks for all the info. So, the path properties are intended to be the logical property name of a property on the model object and then it is up to the retrieve provider to understand that property name and apply the correct filtering logic? There are some ``search`` elements in the FHIR modelinfo that map to FHIR search parameter names (e.g. onset-date) that tricked me into thinking that those path parameters would come across as the actual REST API search parameters names. It is definitely a lot more complicated if the retrieve provider has to map the property name to one or more search strategies. As a simple example, Patient.deceased needs to be treated differently from Condition.onset even though both can be represented as a dateTime. For the patient, the deceased search parameter is a boolean check and for condition, the onset maps to onset-date, onset-age, and onset-info search parameters.

**Comment by JPercival** [comment]
> So, the path properties are intended to be the logical property name of a property on the model object and then it is up to the retrieve provider to understand that property name and apply the correct filtering logic

Correct.

> It is definitely a lot more complicated if the retrieve provider has to map the property name to one or more search strategies.

It is a bit complicated but not quite as bad as that. The cql-translator can detect what the intended type is when there's a choice type (e.g. `Patient.deceased = true` is a Boolean) and only pass the parameters to the retrieve when it's the correct type. For example, `[Condition] C where C.onset = 'childhood'` shouldn't come through as a date filter. Similarly, the `code` filter only supports specific types. If there's a type that's not supported the cql-translator should emit the correct filtering logic in the ELM, which the engine then executes.

**Comment by csandersdev** [comment]
Yeah. I see your point. We do have some guardrails in place because the translator understands types. 

Now that I understand the issue better, I did another pass through the SearchParameterResolver with attention on the normalizePath method where the issue with Condition.onset was introduced. It turns out that the parameter definition is pipe-delimited (e.g. ``Condition.onset.as(dateTime) | Condition.onset.as(Period)``) and the normalizePath method didn't understand that syntax. I created a PR that adds handling to the resolver for pipe-delimited strings that will fix the issues with the Condition.onset and Condition.abatement fields. It isn't the long-term fix for this issue, but would improve the codebase in the interim.
